### PR TITLE
Latest Api changes

### PR DESCRIPTION
**Pull request by osalkk**

### CODE REVIEW

File: terraform/modules/kms_key/variables.tf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+variable "project" {}
```
Comment (JBernC) [inline terraform/modules/kms_key/variables.tf]: descriptions

File: terraform/modules/logs/main.tf [hunk 2]
```diff
@@ -0,0 +1,8 @@
+resource "aws_cloudwatch_log_group" "template_log_group" {
```
Comment (JBernC) [inline terraform/modules/logs/main.tf]: group.tf, stream.tf

File: terraform/modules/logs/variables.tf [hunk 3]
```diff
@@ -0,0 +1,3 @@
+variable "log_group_name" {}
```
Comment (JBernC) [inline terraform/modules/logs/variables.tf]: descriptions

File: terraform/modules/redis/main.tf [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# Redis
+resource "aws_security_group" "sg_elasticache" {
```
Comment (JBernC) [inline terraform/modules/redis/main.tf]: Please split out from main.tf 👍 eg. ``` security.tf, subnet.tf, instance.tf ```

File: terraform/modules/redis/variables.tf [hunk 5]
```diff
@@ -0,0 +1,34 @@
+variable "environment" {}
+
+variable "project" {}
```
Comment (JBernC) [inline terraform/modules/redis/variables.tf]: Descriptions (do you hate me yet???)

File: terraform/modules/s3_encrypted/main.tf [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# data "aws_kms_key" "ob_kms_key" {
+#   key_id = "alias/${var.project}-${var.environment}-kms-key"
+# }
+
+resource "aws_s3_bucket" "bucket" {
```
Comment (JBernC) [inline terraform/modules/s3_encrypted/main.tf]: Is there any way for this module to inherit from the simple S3 bucket and simply add on the serverside encryption?

File: terraform/modules/s3_encrypted/variables.tf [hunk 7]
```diff
@@ -0,0 +1,7 @@
+variable "s3_name" {}
+
+variable "project" {}
+
+variable "environment" {}
```
Comment (JBernC) [inline terraform/modules/s3_encrypted/variables.tf]: descriptions

File: terraform/modules/sns/variable.tf [hunk 8]
```diff
@@ -0,0 +1,15 @@
+variable "sns-topic-policy" {
+  description = "The policy of the particular topic"
+}
+
+variable "name" {
+  description = "Name of the topic"
+}
+
+variable "sqs_arn" {}
```
Comment (JBernC) [inline terraform/modules/sns/variable.tf]: description...

File: terraform/services/anonymiser_service/variables.tf [hunk 9]
```diff
@@ -0,0 +1,15 @@
+variable "region" {}
+
+variable "account" {}
```
Comment (JBernC) [inline terraform/services/anonymiser_service/variables.tf]: descriptions

File: terraform/services/auth_api/iam.tf [hunk 10]
```diff
@@ -0,0 +1,446 @@
+##### DEFINE TL LAMBDA ROLES ######
+resource "aws_iam_role" "iam_role_for_auth_delete_lambda" {
```
Comment (JBernC) [inline terraform/services/auth_api/iam.tf]: Large file, maybe split iam.tf into smaller files specific to the role they are creating aka ```auth_delete_iam.tf```

File: terraform/services/auth_api/lambda.tf [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  service     = "auth"
+  context     = "common"
+  function    = "delete"
+
+  env_vars = {
+    Environment = "${var.environment}"
+  }
+}
+
+module "tl_post_lambda" {
```
Comment (JBernC) [inline terraform/services/auth_api/lambda.tf]: not sure this follows naming convention for clarity.

File: terraform/services/auth_api/ob_api.tf [hunk 12]
```diff
@@ -0,0 +1,77 @@
+# Builds the api gw
+resource "aws_api_gateway_rest_api" "auth_api" {
+  name        = "${var.project}_${var.environment}_auth_api"
+  description = "The API responsible for recieving all open banking and transactional data requests from Users."
+}
+
+# Builds the custom authorizer, only used by get method
+resource "aws_api_gateway_authorizer" "ob_authorizer" {
```
Comment (JBernC) [inline terraform/services/auth_api/ob_api.tf]: should some of this sit in a module???

File: terraform/services/auth_api/tl_reource.tf [hunk 13]
```diff
@@ -0,0 +1,39 @@
+# Use api-resource module to build TL resource, this can be reused for other integrations
```
Comment (JBernC) [inline terraform/services/auth_api/tl_reource.tf]: file name miss-spelled

File: terraform/services/auth_api/variables.tf [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+variable "account" {}
+
+variable "project" {}
+
+variable "runtime" {
+  default = "nodejs6.10"
+}
+
+variable "environment" {}
```
Comment (JBernC) [inline terraform/services/auth_api/variables.tf]: deeeeesssssccccrrriiitttppptttiiiooonnnssss

File: terraform/services/data_consumer_service/variables.tf [hunk 15]
```diff
@@ -0,0 +1,13 @@
+variable "region" {}
```
Comment (JBernC) [inline terraform/services/data_consumer_service/variables.tf]: descriptions

File: terraform/services/data_consumer_service/vpc.tf [hunk 16]
```diff
@@ -0,0 +1,40 @@
+data "aws_subnet_ids" "env_vpc" {
+  vpc_id = "${var.vpc_id}"
+}
+
+data "aws_subnet" "subnet" {
+  count = "${length(data.aws_subnet_ids.env_vpc.ids)}"
+  id    = "${data.aws_subnet_ids.env_vpc.ids[count.index]}"
+}
+
+resource "aws_security_group" "sg_lambda" {
```
Comment (JBernC) [inline terraform/services/data_consumer_service/vpc.tf]: seems fine as a resource but could be a module.

File: terraform/services/deployment/s3.tf [hunk 17]
```diff
@@ -0,0 +1,13 @@
+##### CREATE DEPLOYMENT BUCKET #####
+resource "aws_s3_bucket" "bucket" {
```
Comment (JBernC) [inline terraform/services/deployment/s3.tf]: buckets already exist in modules, no need to make another resource.

File: terraform/services/deployment/variables.tf [hunk 18]
```diff
@@ -0,0 +1,7 @@
+variable "environment" {
+  description = "The environment the firehose+s3+search system sits"
+}
+
+variable "project" {}
```
Comment (JBernC) [inline terraform/services/deployment/variables.tf]: ....

File: terraform/services/dynamodb/dynamodb.tf [hunk 19]
```diff
@@ -0,0 +1,23 @@
+resource "aws_dynamodb_table" "basic-dynamodb-table" {
```
Comment (JBernC) [inline terraform/services/dynamodb/dynamodb.tf]: this exists as a module so shouldn't need another resource.

File: terraform/services/jenkins_cross_account/variables.tf [hunk 20]
```diff
@@ -0,0 +1,9 @@
+variable "bucket_arn" {}
```
Comment (JBernC) [inline terraform/services/jenkins_cross_account/variables.tf]: descriptions

File: terraform/services/stats/firehose.tf [hunk 21]
```diff
@@ -0,0 +1,29 @@
+##### CREATE ANALYTICS STREAM #####
+resource "aws_kinesis_firehose_delivery_stream" "firehose_stream" {
```
Comment (JBernC) [inline terraform/services/stats/firehose.tf]: Is there not a firehose module.

File: terraform/services/auth_api/lambda.tf [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+  service     = "auth"
+  context     = "tl_access"
+  function    = "post"
+
+  env_vars = {
+    TRANSACTION_STREAM_NAME   = "${var.transactions_stream_name}"
+    TRANSACTION_PARTITION_KEY = "transaction-partition-key"
+    STATS_PARTITION_KEY       = "stats-kinesis-partition-key"
+    STATS_STREAM_NAME         = "${var.stats_stream_name}"
+    TL_CLIENT_ID              = "clearscore-ra5i"
```
Comment (JBernC) [inline terraform/services/auth_api/lambda.tf]: Don't think the post lambda needs client id or cclient secret. Also Hs256 cert shouldn't be external (although leave as is for now)

File: terraform/services/data_consumer_service/lambda.tf [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+  source             = "../../modules/api_lambda_vpc"
+  handler            = "handler"
+  runtime            = "nodejs6.10"
+  role               = "${aws_iam_role.iam_role_for_consumer_lambda.arn}"
+  bucket             = "${var.lambda_source_bucket}"
+  environment        = "${var.environment}"
+  project            = "${var.project}"
+  service            = "plugin"
+  context            = "cache"
+  function           = "data"
```
Comment (JBernC) [inline terraform/services/data_consumer_service/lambda.tf]: should be transactions not data

File: environments/development.tfvars [hunk 24]
```diff
@@ -0,0 +1,34 @@
+// Development environment variables
+country_code = "gb"
+
+organization = "clearscore"
+
+project = "openbanking"
+
+region = "eu-west-1"
```
[the remark this answers is not in the document]
Comment (osalkk) [reply environments/development.tfvars]: Don't think so, it's better to use in variables so we don't have to bother for each environment.

File: terraform/modules/api-method-auth/main.tf [hunk 25]
```diff
@@ -0,0 +1,48 @@
+resource "aws_api_gateway_method" "request_method" {
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api-method-auth/main.tf]: I think for the modules, we should leave them as is. If we split them we will just create a file for each resource. For the services, I agree with you but for modules using main, variables and outputs are fine according to me.

File: terraform/modules/api_lambda/main.tf [hunk 26]
```diff
@@ -0,0 +1,21 @@
+data "aws_s3_bucket_object" "zip" {
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api_lambda/main.tf]: unnecessary as you said

File: terraform/modules/api_lambda/variables.tf [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+variable "handler" {
+  description = "The handler name of the lambda (a function defined in your lambda)"
+  default     = "handler"
+}
+
+variable "role" {
+  description = "IAM role attached to the Lambda Function (ARN)"
+}
+
+variable "project" {}
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api_lambda/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/modules/api_lambda_vpc/main.tf [hunk 28]
```diff
@@ -0,0 +1,26 @@
+data "aws_s3_bucket_object" "zip" {
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api_lambda_vpc/main.tf]: unnecessary again

File: terraform/modules/api-method-auth/variables.tf [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+
+variable "account" {
+  description = "The AWS account ID"
+}
+
+variable "type" {
+  description = "type of integration lambda has with api"
+}
+
+variable "authorization" {}
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api-method-auth/variables.tf]: added

File: terraform/modules/api-method/main.tf [hunk 30]
```diff
@@ -0,0 +1,45 @@
+resource "aws_api_gateway_method" "request_method" {
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api-method/main.tf]: We agreed not to split the modules and also connect has the same structure for the modules as well

File: terraform/modules/api_lambda/main.tf [hunk 31]
```diff
@@ -0,0 +1,21 @@
...
+resource "aws_lambda_function" "lambda" {
+  s3_bucket     = "${data.aws_s3_bucket_object.zip.bucket}"
+  s3_key        = "${data.aws_s3_bucket_object.zip.key}"
+  function_name = "ob_${var.service}_${var.context}_${var.function}_${var.environment}"
+
+  role    = "${var.role}"
+  handler = "index.${var.handler}"
+  runtime = "${var.runtime}"
+
+  memory_size = "128"
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api_lambda/main.tf]: Done

File: terraform/modules/api_lambda/variables.tf (hunk 27 above)
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api_lambda/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/modules/api_lambda_vpc/main.tf [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+  function_name = "ob_${var.service}_${var.context}_${var.function}_${var.environment}"
+
+  role    = "${var.role}"
+  handler = "index.${var.handler}"
+  runtime = "${var.runtime}"
+
+  memory_size = "128"
+  timeout     = "60"
+
+  vpc_config {
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api_lambda_vpc/main.tf]: nope, vpc is the attribute for the lambda, not a separate module

File: terraform/modules/api_lambda_vpc/variables.tf [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+variable "handler" {
+  description = "The handler name of the lambda (a function defined in your lambda)"
+  default     = "handler"
+}
+
+variable "role" {
+  description = "IAM role attached to the Lambda Function (ARN)"
+}
+
+variable "project" {}
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api_lambda_vpc/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/modules/kinesis/variables.tf [hunk 34]
```diff
@@ -0,0 +1,22 @@
+variable "name" {}
+
+variable "environment" {}
```
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/kinesis/variables.tf]: description for the name is added

File: terraform/modules/kms_key/variables.tf (hunk 1 above)
Comment (osalkk) [reply terraform/modules/kms_key/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/modules/logs/main.tf (hunk 2 above)
Comment (osalkk) [reply terraform/modules/logs/main.tf]: I never used that module, what is it for?

File: terraform/modules/logs/variables.tf (hunk 3 above)
Comment (osalkk) [reply terraform/modules/logs/variables.tf]: I never used that module, what is it for?

File: terraform/modules/redis/main.tf (hunk 4 above)
Comment (osalkk) [reply terraform/modules/redis/main.tf]: We agreed not to split the modules

File: terraform/modules/redis/variables.tf (hunk 5 above)
Comment (osalkk) [reply terraform/modules/redis/variables.tf]: Do we really need description for project, environment, country_code? ( Yes I hate you 🗡 )

File: terraform/modules/s3_encrypted/variables.tf (hunk 7 above)
Comment (osalkk) [reply terraform/modules/s3_encrypted/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/modules/sns/variable.tf (hunk 8 above)
Comment (osalkk) [reply terraform/modules/sns/variable.tf]: We still don't know if we are going to use the sqs, so let's leave it for now

File: terraform/services/anonymiser_service/variables.tf (hunk 9 above)
Comment (osalkk) [reply terraform/services/anonymiser_service/variables.tf]: Do we really need description for project, environment, country_code,region, account?

File: terraform/services/auth_api/ob_api.tf (hunk 12 above)
Comment (osalkk) [reply terraform/services/auth_api/ob_api.tf]: No, we only use on authorizer.

File: terraform/services/auth_api/lambda.tf (hunk 11 above)
Comment (osalkk) [reply terraform/services/auth_api/lambda.tf]: It follows /tl
                  |_ Post --> lambda

File: terraform/services/auth_api/variables.tf (hunk 14 above)
Comment (osalkk) [reply terraform/services/auth_api/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/services/data_consumer_service/variables.tf (hunk 15 above)
Comment (osalkk) [reply terraform/services/data_consumer_service/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/services/data_consumer_service/vpc.tf (hunk 16 above)
Comment (osalkk) [reply terraform/services/data_consumer_service/vpc.tf]: Let's discuss this - To do

File: terraform/services/deployment/s3.tf (hunk 17 above)
Comment (osalkk) [reply terraform/services/deployment/s3.tf]: done

File: terraform/services/deployment/variables.tf (hunk 18 above)
Comment (osalkk) [reply terraform/services/deployment/variables.tf]: Do we really need description for project, environment, country_code?

File: terraform/services/jenkins_cross_account/variables.tf (hunk 20 above)
Comment (osalkk) [reply terraform/services/jenkins_cross_account/variables.tf]: done

File: terraform/services/stats/firehose.tf (hunk 21 above)
Comment (osalkk) [reply terraform/services/stats/firehose.tf]: no, there is no firehose in London as well 😄

File: terraform/services/auth_api/lambda.tf (hunk 22 above)
Comment (osalkk) [reply terraform/services/auth_api/lambda.tf]: No idea about that let's discuss this

File: terraform/services/data_consumer_service/lambda.tf (hunk 23 above)
Comment (osalkk) [reply terraform/services/data_consumer_service/lambda.tf]: It is data

File: terraform/modules/api-method/main.tf (hunk 30 above)
[the remark this answers is not in the document]
Comment (osalkk) [reply terraform/modules/api-method/main.tf]: Didn't get that, let's discuss this

**Pull request merged**